### PR TITLE
Add support for non-independent sibling modules in layer contracts

### PR DESCRIPTION
**Pull request by Peter554**
This PR contains the changes needed within import-linter to support https://github.com/seddonym/import-linter/issues/208.

This continues the prerequisite work in grimp to support this feature https://github.com/seddonym/grimp/pull/136.

### CODE REVIEW

File: pyproject.toml
Comment (Peter554) [inline pyproject.toml:16]: This commit needs to be removed before merge! ⚠️ 

I've tested locally against this version of grimp, all seems to work well.

**Comment by Peter554** [comment]
hi @seddonym 👋 here's a PR for the import-linter changes. it seems to work well for me when testing against the latest version of grimp. I'm not sure if the CI pipeline is going to pass here though - not sure if I set the grimp dependency correctly (locally I just created a virtualenv and ran `pip install "git+https://github.com/seddonym/grimp.git@88d49122dd66fe6d092a69a5e8a302e667e908d1"`).

File: src/importlinter/contracts/layers.py
Comment (Peter554) [inline src/importlinter/contracts/layers.py:35]: Not sure exactly what error we should raise here, any suggestions?

File: pyproject.toml
Comment (Peter554) [reply pyproject.toml]: or more accurately, the new version of grimp needs to be published and set here.

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:30]: Something to be aware of when making changes to the fields is it is possible that someone is using them for a [custom contract](https://import-linter.readthedocs.io/en/stable/custom_contract_types.html). So this could be a breaking change, in theory.

In practice I'm not sure how much custom contracts are used, but ideally changes should be backwards-compatible. I'm open to relaxing this in certain cases though. Probably one for discussion.

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:34]: Could we define `|` and `:` as constants to keep it DRY?

Comment (seddonym) [inline src/importlinter/contracts/layers.py:34]: This method is getting a little long, I wonder if we could break out the logic into smaller methods? Not a blocker.

File: src/importlinter/contracts/layers.py
Comment (seddonym) [reply src/importlinter/contracts/layers.py:35]: It should raise `importlinter.domain.fields.ValidationError`.

File: tests/unit/contracts/test_layers.py
Comment (seddonym) [inline tests/unit/contracts/test_layers.py:38]: Could we have a test for the invalid mixture of two kinds of dividers?

File: tests/unit/contracts/test_layers.py
Comment (seddonym) [inline tests/unit/contracts/test_layers.py:172]: Maybe this test is doing too much? For a start it is called `test_illegal_imports`, but one of the cases is for no illegal imports.

I think I would lean towards having an entirely separate test case, or at least renaming the test to be accurate. Thoughts?

**Review by seddonym — CHANGES_REQUESTED** [review]
This is in good shape...I've left a couple of small comments.

But I have been thinking a bit more about the API for our recent work on Grimp. It's been niggling me that the argument to `find_illegal_dependencies_for_layers` is called `layers` but we give it a sequence of levels, which contain the layers.

I think we should consider making the concept of 'level' that contains 'layers' actually be a 'layer' that contains something else ('modules' perhaps). It just feels conceptually cleaner. At the very least that's just a rename of `grimp.Level` -> `grimp.Layer` - though it would at some point be nice to tighten up the terminology throughout Grimp, including in the rust code.

Interested to hear your thoughts, including how it would impact this PR.

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: Thanks, changed it

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: Good point. I would guess that custom contracts aren't used so much, and when they are used I'm not sure if people are importing from the existing contracts to extend them or just starting from scratch. For me this breaking change is okay - if someone really wants the old `LayerField` then they can just copy-paste those small number of lines from the old version (very low effort). It's also only a linting tool, so will only break a CI pipeline at worst (not a real running app). What do you think?

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: Sure, done

File: tests/unit/contracts/test_layers.py
Comment (Peter554) [reply tests/unit/contracts/test_layers.py:49]: Yes good idea, added

File: tests/unit/contracts/test_layers.py
Comment (Peter554) [reply tests/unit/contracts/test_layers.py:197]: I quite like to use the same test so that we can sure the setup is the same, so that we can be sure that there is no other unintended difference between the test cases. For now I've renamed the test to `test_imports_between_sibling_modules`.

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:239]: Could we rename `declared_modules` to  `declared_module_tails` (now we have a name for it)? And, lower down, could we rename `undotted_module` to `module_tail`?

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:31]: Could we rename to `is_independent` and default this to `True`?

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:30]: 🤔 Might be worth making this a frozenset, so `Layer` becomes hashable - that allows us to add it to sets. On the other hand it's a bit more awkward to notate frozensets so I'll leave it up to you, we can always adjust later.

File: tests/unit/contracts/test_layers.py
Comment (seddonym) [inline tests/unit/contracts/test_layers.py:21]: These can be skipped if independent is True by default.

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:55]: Would be clearer to say "Layer cannot have a mixture of independent and non-independent elements."?

Also, can the error message have a full stop at the end of the sentence?

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:37]: It feels like it would be simpler if we broke this into two functions calls, one that tells us whether the layer is independent, and one that splits the elements. The splitter could be agnostic of what the delimiter is and not care if there are two different kinds.

Do you agree?

Comment (seddonym) [inline src/importlinter/contracts/layers.py:37]: Could we rename `independent` to `is_independent`, to clarify it's a boolean?

File: docs/contract_types.rst
Comment (seddonym) [inline docs/contract_types.rst:194]: For consistency maybe we should make this `Contract with sibling modules (independent)`?

File: docs/contract_types.rst
Comment (seddonym) [inline docs/contract_types.rst:214]: Worth adding this sentence, like we have for the independent example?

```
Unlike the previous example, in this case ``medium_a``, ``medium_b`` and ``medium_c``
are not enforced as independent, so are allowed to import from each other.
```

File: CHANGELOG.rst
Comment (seddonym) [inline CHANGELOG.rst:7]: Maybe better: "Add support for non-independent sibling modules in layer contracts."

File: tests/unit/contracts/test_layers.py
Comment (seddonym) [inline tests/unit/contracts/test_layers.py:196]: This line can be removed.

File: tests/unit/contracts/test_layers.py
Comment (seddonym) [reply tests/unit/contracts/test_layers.py:197]: Ok, no problem!

File: src/importlinter/contracts/layers.py
Comment (seddonym) [inline src/importlinter/contracts/layers.py:29]: This is a potentially breaking change, if a developer has built a custom contract type which uses either `LayerField` or the `Layer` class. That being said, I do not know if any such developer actually exists!

I think it's a good change though. I'm inclined to leave as-is, but include a note in the CHANGELOG and release this with a major version bump.

Suggested text for CHANGELOG:

```
- In ``importlinter.contracts.layers``, ``Layer`` and ``LayerField`` 
  have changed their API slightly. This could impact custom
  contract types depending on these classes. 
```

**Review by seddonym — CHANGES_REQUESTED** [review]
Great stuff.

I've requested a few changes but hopefully these are relatively easy to action. In the meantime I'm going to get Grimp ready for a release, so you should be able to drop the pinned version before merging.

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py:23]: Added

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: Yes, makes more sense. Updated.

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: Sure, gone with `_grimpify_layers`

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: yep 👍

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: done (although in grimp we called it just `independent`)

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py:37]: I've left it as `set` for now, as you say it's pretty easy to adjust later if needed

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py]: I've split this into two separate method calls now

File: src/importlinter/contracts/layers.py
Comment (Peter554) [reply src/importlinter/contracts/layers.py:36]: Yes, added this line to the changelog

File: src/importlinter/contracts/layers.py
Comment (seddonym) [reply src/importlinter/contracts/layers.py]: Hmm, that's a good point. I'm in the habit of naming booleans like this, but perhaps it's better to have it the same as Grimp. I'll leave it up to you whether you want to change it back to just `independent`, I don't mind too much.

**Review by seddonym — COMMENTED** [review]
This is great, thank you. I'll hold off approving until it's using the released version of Grimp - I'm working on that now.

**Comment by seddonym** [comment]
Grimp 3.2 is now released so you can update the dependencies.

**Comment by Peter554** [comment]
Thanks - updated the grimp version here now, so I think we're ready to go.

**Review by seddonym — APPROVED** [review]

**Pull request merged**
